itest/lnd_channel_funding_utxo_selection_test.go

```diff
@@ -0,0 +1,309 @@
+			chanOpenShouldFail: true,
+			feeRate:            15,
+			expectedErrStr: "output amount(0.00000174 BTC) after " +
+				"subtracting fees(0.00001826 BTC) below dust " +
+				"limit(0.0000033 BTC)",
+		},
+		// Selected coins don't cover the minimum channel size.
+		{
+			name: "fundmax, local amount < min chan " +
+				"size",
+			initialCoins:       []btcutil.Amount{18_000},
+			selectedCoins:      []btcutil.Amount{18_000},
+			feeRate:            1,
+			chanOpenShouldFail: true,
+			expectedErrStr: "available funds(0.00017877 BTC) " +
+				"below the minimum amount(0.0002 BTC)",
+		},
+		// The local amount exceeds the value of the selected coins.
+		{
+			name: "selected, local amount > " +
+				"selected amount",
+			initialCoins:       []btcutil.Amount{100_000, 50_000},
+			selectedCoins:      []btcutil.Amount{100_000},
+			localAmt:           btcutil.Amount(210_337),
+			chanOpenShouldFail: true,
+			expectedErrStr: "not enough witness outputs to " +
+				"create funding transaction, need 0.00210337 " +
+				"BTC only have 0.001 BTC  available",
+		},
+		// We are spending two selected coins partially out of three
+		// available in the wallet and expect a change output and the
+		// unselected coin as remaining wallet balance.
+		{
+			name: "selected, local amount > " +
+				"min chan size",
+			initialCoins: []btcutil.Amount{
+				200_000, 50_000, 100_000,
+			},
+			selectedCoins: []btcutil.Amount{
+				200_000, 100_000,
+			},
+			localAmt:        btcutil.Amount(250_000),
+			expectedBalance: btcutil.Amount(250_000),
+			remainingWalletBalance: btcutil.Amount(350_000) -
+				btcutil.Amount(250_000) -
+				fundingFee(2, true),
+		},
+		// We are spending the entirety of two selected coins out of
+		// three available in the wallet and expect no change output and
+		// the unselected coin as remaining wallet balance.
+		{
+			name: "fundmax, local amount > min " +
+				"chan size",
+			initialCoins: []btcutil.Amount{
+				200_000, 100_000, 50_000,
+			},
+			selectedCoins: []btcutil.Amount{
+				200_000, 50_000,
+			},
+			expectedBalance: btcutil.Amount(200_000) +
+				btcutil.Amount(50_000) -
+				fundingFee(2, false),
+			remainingWalletBalance: btcutil.Amount(100_000),
+		},
+		// We fund an anchor channel with a single coin and just keep
+		// enough funds in the wallet to cover for the anchor reserve.
+		{
+			name: "fundmax, sufficient reserve",
+			initialCoins: []btcutil.Amount{
+				200_000, 10_000,
+			},
+			selectedCoins:  []btcutil.Amount{200_000},
+			commitmentType: lnrpc.CommitmentType_ANCHORS,
+			expectedBalance: btcutil.Amount(200_000) -
+				fundingFee(1, false),
+			remainingWalletBalance: btcutil.Amount(10_000),
+		},
+		// We fund an anchor channel with a single coin and don't
+		// provide enough funds in the wallet to cover for the reserve.
+		// Hence the channel funding amount is decreased and a change
+		// output created in order to satisfy the reserve requirement.
+		{
```

Comment:
adding a case where:
- local amt is specified and
- part of the reserve needs to come from the specified utxos

would be good. my intuition would a failure since we won't be able to respect the user's request local amt

Reply:
Thanks for the suggestion. I have added some tests and will commit them soon. Your intuition is right if the user selection of utxos and local_amt would result in change that, combined with the unselected, remaining wallet balance would leave insufficient reserves in the wallet.